### ISSUE TITLE
Considerations on interoperability and evolution

### ISSUE THREAD

**Issue by gouttegd** [mapping-commons/sssom#325]
What are the plans, if any, to ensure the SSSOM TSV format remains interoperable

* across versions (a file produced with an implementation conforming to version _X_ can be used by an implementation conforming to version _X+1_),
* across implementations (a file produced by implementation _X_ — e.g. `sssom-py` — can be used another implementation _Y_ — e.g. `sssom-java`)?

Currently, this interoperability goal is explicitly _not_ sought after, as indicated both by the lower-than-one major version number (which, if we assume the project is using [semantic versioning](https://semver.org/), means that _anything_ can change _at any time_) and by the explicit warning on the top-level README: _Note that SSSOM is currently under development and subject to change._

With the SSSOM format being seemingly more and more used in the wild, I believe it is time to consider committing to some form of long-term stability of the format, and/or design ways to make the format evolve while preserving some basic interoperability across versions and implementations.

The following is a random set of ideas that could be explored. Feel free to discuss them, refute them, and add more.

## Defining a “core” set of metadata that will never change

We could select a handful of the most important mapping metadata and promote them to a “core set” that would be guaranteed never to change in any future evolution. This is similar to the “[minimal spec](https://github.com/mapping-commons/sssom/issues/260)” idea, though it could probably include slightly more metadata slots than the four mentioned in that ticket.

The principle here is that potential users could be confident that, no matter how the format evolves over time, as long as they only use the “core set” their files would always remain exploitable by any version of any conforming implementation. For users who need metadata from the “non-core set”, the situation would be the same as it is now for the entire standard: they would need to watch carefully the evolution of the standard to avoid being surprised by a breaking change.

For implementations, this would mean that they should only be strict when parsing the “core” metadata (at least by default — of course they can choose to allow users to specify a different behaviour). If they encounter a metadata slot they don’t recognise (because it’s an addition from a newer version of the spec), or a slot whose format has changed (e.g. because of a change such as the one envisioned [here](https://github.com/mapping-commons/sssom/issues/323), they may log a warning but should not fail altogether to parse the file.

## Allow each set to declare its own “must-understand” slots

This can be seen as a variation of the “core set” idea. Here, instead of having a fixed list of “core” metadata slots, the creators of a mapping set could define their own list of the slots they consider as critical.

For example, considering the following set (and assuming that the `similarity_threshold` slot, proposed [here](https://github.com/mapping-commons/sssom/issues/296), and the `mapping_chain_intermediate` slot, proposed [here](https://github.com/mapping-commons/sssom/issues/231), have been added to a later version of the spec):

```
# must-understand:
#   - mapping_chain_intermediate
subject_id	predicate_id	predicate_modifier	object_id	mapping_justification	similarity_threshold	mapping_chain_intermediate
EXA:1234	skos:exactMatch	Not	EXB:5678	semapv:ManualMappingCuration	0.8	EXC:4321
```

An implementation trying to read that file would first check the list of the “must-understand” slots for any slot that it does not recognise, and should flatly reject the file if it does contain such a slot.

So, an implementation up-to-date with the latest version of the spec (and thus, which supports both `mapping_chain_intermediate` and `similarity_threshold`) would parse the file without any issue. An implementation that for whatever reason does _not_ recognise `mapping_chain_intermediate` (maybe because it has not been updated to catch up with the latest version of the spec yet) should immediately fail with an error.

## Adding a slot for the spec version

We could add a simple `sssom-version` metadata slot at the mapping set level, to indicate the version of the spec this set is conforming to. Ideally that slot would be required to be the very first slot listed in the metadata block, so that a parser could figure out immediately whether the file it is trying to read is using a version it supports.

It would be up to the implementations to decide whether they want to support several versions at the same time or not.

## Versioning the format in addition to the spec

Regardless of whether we add a versioning slot, it could be useful to introduce a version number for the _file format_, distinct from the version number of the _specification_. Not all changes to the specification have an impact on the file format, so tracking the evolution of the spec separately from the evolution of the format would make sense.

**Comment by matentzn** [comment]
This is of great concern to me as well, and related to https://github.com/mapping-commons/sssom/issues/189.

I do not know the best answer to any of your questions, and I am happy to yield to advice by more experienced people in standards management. Here is my rough sense:

> Allow each set to declare its own “must-understand” slots

Seem to be too complicated to implement in practice.

> Adding a slot for the spec version

I think this is a very good idea. Even if you cant open a specific version of an (older) file, this could give a hand which tool sssom-py, sssom-java version could deal with this. I would support such a slot potentially even as "mandatory".

> Core slots

This is my cautious weak position:

1. I think the idea is good. We have a notion of core properties in sssom-py which are all required properties, plus predicate_modifier. We could stableise these right now. But how do we communicate this?
2. We should probably agree that file format breaking changes should always require a new _major_ version upgrade, while format changes that do not affect parsing (metadata etc) could be dealt with by minor version updates. This means that all file formats are intrinsically tied to a major release of SSSOM, say 1.0, 2.0 etc.

**Comment by matentzn** [comment]
I think we are on the same page regarding where we are going with this. We need a little to substantiate this understanding:

> Currently, the two are completely intertwined, because the file format is basically a direct serialisation of the data model.

As far as I understand, this is only really true for the YAML serialisation, which we have not promoted much at all; I think the "mapping" from the data model to their serialisations is done via [transformation procedures](https://linkml.io/linkml/data/conversion.html#linkml-convert). So there is actually a double risk: not only can the changes to the datamodel directly affect json and rdf serialisations, but also changes to the transformation procedures. So I now think we should probably do as you say:

1. Define serialisation versions
2. Tie them to specific versions of linkml transform (this has the ugly ugly ugly consequence that we need to pin any specific version of sssom-py to a specific version of linkml-runtime) and sssom-schema
3. Add unit tests for each release when updating linkml (deliberately) to ensure the serialisation does not change.

Ugh.. You are shining some torchlight into the underdark @gouttegd..

**Comment by gouttegd** [comment]
> As far as I understand, this is only really true for the YAML serialisation

It is also partially the case for the SSSOM-TSV format, with `sssom-py`.

When _writing_ a SSSOM-TSV file, `sssom-py` writes the metadata block by basically dumping the metadata dictionary as it is (`yaml.safe_dump(meta)`) and writes the TSV part by basically dumping the Pandas data frame (`msdf.df.to_csv()`). So what’s written _is_ a direct serialisation of the SSSOM data model, so there’s no room for the serialisation format and the data model to diverge.

The good news is that when _reading_ though, `sssom-py` first loads the metadata block and the TSV rows as they are, and _then_ compares them with what it expects from the SSSOM schema – meaning that here it would be possible from the serialisation format to diverge from the data model, because the parser could perform any necessary “translation” (for example, replace `owl class` by `owl:Class`, if we take again the example of #323).

> I think the "mapping" from the data model to their serialisations is done via [transformation procedures](https://linkml.io/linkml/data/conversion.html#linkml-convert).

Unless I missed something, I don’t think so, at least not for SSSOM-TSV and not in the current version of `sssom-py`.

**Comment by gouttegd** [comment]
> Tie them to specific versions of linkml transform

Please don’t. The serialisation format should be specified on its own, without any reference to any LinkML stuff.

Reference to LinkML is precisely the problem with the JSON serialisation format (#321). The “specification” just says, “the JSON format is whatever LinkML’s serialisation procedures do.” In effect this makes this format impossible to implement in any language that does not have decent LinkML support (so, any language other than Python¹).

---
¹Yeah, I know, [“LinkML is *definitely not* only for Python developers”](https://linkml.io/linkml/faq/general.html#is-linkml-only-for-python-developers)… Believe that if you want. As someone who has been implementing two LinkML-defined projects in Java,  I can tell you the support for other languages in LinkML is symbolic at most. Make the SSSOM serialisation format entirely dependent on the LinkML runtime and the only result will be that nobody will ever try to support SSSOM in any other language than Python — which would mean for example no more SSSOM plugin for ROBOT.

**Comment by matentzn** [comment]
> Unless I missed something, I don’t think so, at least not for SSSOM-TSV and not in the current version of sssom-py.

sssom TSV is indeed different, for all other serialisations though, it is true.

As always, all you say is sane and we should go this way. What is the best way to document the file format independently of the schema though? I thought this is precisely what the LinkML JSON Schema export was for.

Before we start with anything, we should probably build a library of examples and, similar to ROBOT, build a test suite that ensures that with every release, we do not break serialisation expectations..

**Comment by gouttegd** [comment]
> As always, all you say is sane and we should go this way.

Err, which way? My mind is _not_ set on any way in particular (sorry if I gave the impression it was), I was merely throwing ideas. ^^'

> What is the best way to document the file format independently of the schema though?

The documentation/specification does not need to be completely “independent of the schema”. What it must be (and that’s the only aspect on which I do have a firm opinion) is to be independent of the LinkML _runtime_ – that is, it must not do what the description of the JSON format is doing. The SSSOM-TSV format _is_ currently described independently of the runtime, and it should remain so.

In fact, the way the SSSOM-TSV format is currently specified is pretty good, I think. It’s a textual description that refers to the LinkML schema when needed (e.g., to avoid manually listing all the metadata fields or the meaning of some enumeration values).

What needs to be decided is if we want to allow the SSSOM-TSV format to deviate from the schema for backwards-compatibility reasons, so that the format (or rather, its parsers) can act as a compatibility layer that shields users from breaking changes in the schema.

If we want do to that, it can be done simply by adding a section in the spec that would say something like

> ## Compatibility with previous versions
>
> In addition to the columns described in the metadata table mentioned above, a compliant parser MUST be ready to accept the following columns from previous versions of the specification:
>
> ### `match_type`
> This column was used to indicate how a mapping had been asserted in versions prior to 0.9.1. If it is present, the parser MUST silently convert it to a `mapping_justification` slot using the following table:
>
> * `Lexical` -> `semapv:LexicalMatching`
> * `HumanCurated` -> `semapv:ManualMappingCuration`
> [etc.]
>
> ### `match_term_type`
> This column was used to describe the type of entities being matched. It was replaced in SSSOM 0.9.1 by two distinct columns `subject_type` and `object_type`. If it present, the parser MUST convert it to those two columns using the following table:
>
> * `ClassMatch` -> `owl class`
> * `ConceptMatch` -> `skos concept`

Of course, this is only an example. These changes occurred before version 1.0, so it would be fine _not_ to guarantee backwards compatibility in this case. But this illustrates how we could do it if we need to introduce breaking changes after version 1.0.

**Comment by gouttegd** [comment]
> What needs to be decided is if we want to allow the SSSOM-TSV format to deviate from the schema for backwards-compatibility reasons

If we do _not_ want to do that (that is, we want the SSSOM-TSV format to always follow closely the schema, as it currently does), then I think that adding a version number to the metadata becomes necessary.

**Comment by gouttegd** [comment]
Trying to summarise the options:

1) Backwards compatibility is out of scope for the specification.

We decide that the spec simply does not care about backwards compatibility. The spec is only ever about the latest (current) version. We update the schema as we see fit. At most, we promise, once version 1.0 is out, to break things as rarely as possible. But backwards compatibility is entirely left at the discretion of the implementations: some may only support the latest version of the spec, other may try to do their best to support previous versions as well.

2) Backwards compatibility is out of scope, but with a version number.

Basically same as option 1), but we do add a `sssom_version` field to the dataset metadata so that implementations can at least immediately know whether the file they are trying to read is using a version of the model that they support (instead of figuring that out the hard way, for example upon encountering a metadata slot that they do not recognise). It is again left at the discretion of the implementations whether they want to support older versions and how they do so.

Both options 1 and 2 do not require the SSSOM-TSV format to deviate from the data model.

3) Backwards compatibility is a spec feature

The spec explicitly a) allows, b) encourages, or c) mandates (pick one option) implementations to support older versions. The SSSOM-TSV format is allowed to deviate from the latest data model. Every new version of the spec includes some backwards compatibility notes describing what has changed, what old metadata slots are no longer present in the data model but may still be found in SSSOM-TSV files, and how to deal with such slots (e.g. how to convert them to the current data model).

A `sssom_version` metadata slot, as in option 2, would also be nice here, though not indispensable.

I am inclined towards option 3b, personally.

**Comment by matentzn** [comment]
> Err, which way? My mind is not set on any way in particular (sorry if I gave the impression it was), I was merely throwing ideas. ^^'

Sorry I should have referred to your quote on _not tying_ to a specific linkml runtime version. :D 

> What needs to be decided is if we want to allow the SSSOM-TSV format to deviate from the schema for backwards-compatibility reasons, so that the format (or rather, its parsers) can act as a compatibility layer that shields users from breaking changes in the schema.
> If we want do to that, it can be done simply by adding a section in the spec that would say something like

I think we could do something like that. Of course, this approach will also have certain limits (for example, if the free field (not enum) changes from string to entityref datatype, and you don't know all the possible values like in the `match_type` case))

I think we can go this way.

> If we do not want to do that (that is, we want the SSSOM-TSV format to always follow closely the schema, as it currently does), then I think that adding a version number to the metadata becomes necessary

I think we should do that anyways. Making this an optional property that defaults to the latest version.

> Trying to summarise the options:

Great summary!

Can you play the following through: what exactly would happen if an uncontrolled field (no fixed values) is changed from type string to type EntityRef? This happened twice in the past, with `subject_source` and `mapping_justification` (aka `match_type`). Are you saying that when writing file where a value is still a string, the reference writers 

1. Will try to repair the value silently if not ambiguous
2. Will be forced to write the value as it was (maybe printing a warning)

I am not against encouraging implementors to support older versions, but I fear that this will create a big communication overhead (with resources that have their own parsers). Every more small variants of the file formats will make the parser requirements ever more complex. Validation methods will become stranger too, as the parsers required for validation are very much tied to the schema at the moment (linkml validate). I am personally more inclined towards (2), so far.

**Comment by gouttegd** [comment]
> this approach will also have certain limits (for example, if the free field (not enum) changes from string to entityref datatype, and you don't know all the possible values like in the match_type case))

Yes. Depending on the breaking changes we introduce, there will not always be a straightforward way to ensure backwards compatibility.

> what exactly would happen if an uncontrolled field (no fixed values) is changed from type string to type EntityRef?

It will happen _whatever we’ll decide should be happening in that case_, nothing more, nothing less. :)

There is no pre-defined answer. Every time we introduce a breaking change, we’ll have to consider what could/should be done.

That may depend on many things, such as:

* Is the field that we are breaking an “important” field? For example, when `match_type`  was replaced by `mapping_justification`, this was touching a very important field. For such a field, we should probably go to greater length to ensure compatibility than for a more “minor” field.
* Do we have any idea of how widely the field has been used in the wild?
* For an uncontrolled field, is there a way we could get a glimpse of _how_ it has been used (what type of values people put in it)?
* When the field had been used, _what_ has it been used for?

What should probably be done to get some answers to at least some of these questions is to advertise beforehand our intention to change the field (e.g. on a GitHub ticket, with call to comments on Slack and any other channel you can think of).

And then, we decide: if we go ahead with the changes, and if yes, how do we suggest / recommend / mandate that implementations deal with the change. Options may include:

1. Don’t even try to deal with it. It’s only a minor field, just ignore it. Print a warning if you want, but then forget it and process the rest of the mapping set normally.
2. Don’t even try to deal with it. It’s a field that is too important for you to try to guess how to convert it to the new model. Just abort immediately if you come across the old field. Leave it to the user to amend her mapping set with the new field. Point her to this very helpful page we have prepared in which we explain the issue.
3. Use the table below that lists some commonly used values for this field along with the EntityRef that should now be used instead. If the mapping set your parsing contains values that are not in this list, use the special value `semapv:IHaveNoIdeaWhatThisIsAbout`, print a warning, but keep parsing the mapping set.

> Validation methods will become stranger too, as the parsers required for validation are very much tied to the schema at the moment (linkml validate)

That is not a problem as long as validation happens _after_ any compatibility preprocessing.

For example, if a mapping set is still using `match_type`, the compatibility layer in the parser will convert that to `mapping_justification`, and then the validator will only ever see `mapping_justification`. The validator needs not be changed in any way; it can remain aware only of the latest version of the schema.

>  I am personally more inclined towards (2)

The main difference between (2) and (3a) is that, while in both cases backwards compatibility is entirely optional, in (3a) at least the spec provides informations about how to do it.

Basically, option (2) is: “This document only describes the latest version of the spec. You MUST support that version as described here. It’s entirely up to you whether you want to support older versions, but if you do, it’s up to you to dig up older versions of this document, figure out what has changed, and coming up with a way to deal with those changes. Have fun!”

Option (3a) is: “This document describes the latest version of the spec. You MUST support that version as described here. You MAY support older versions if you want, it’s up to you. If you want to do that, please refer to the COMPATIBILITY WITH OLDER VERSIONS section below for details of what has changed in the last version and hints on how to deal with older versions.”

One issue with option (2) is that if you leave each implementations come up with their own solutions, well… they will _each come up with their own solutions_, which may very well be completely different from one implementation to the next. One implementation will silently ignore that old field, another one will reject it outright, yet another one will silently convert it in one way, and yet another will convert it in another slightly incompatible way.

**Comment by gouttegd** [comment]
> Support for older versions does not need to be a “all-or-nothing” choice. The spec can offer the choice between MAY / SHOULD / MUST for every single compatibility issue.

This also means that “supporting older versions” does not need to mean “support for _all_ older versions”. We can fine-tune the expected level of support for each version.

For example, if we make a SSSOM version 1.1, it would be perfectly reasonable, in my opinion, to ask that implementations compliant with that version SHOULD (possibly even MUST) also support version 1.0, because presumably there wouldn’t be much differences between 1.0 and 1.1.

Later, when we get to SSSOM 2.0, it would also be perfectly reasonable to say that implementations compliant with that version MAY support any 1.x version (meaning that they can do it if they want, but they are _not_ expected to).

**Comment by matentzn** [comment]
I cannot tell you enough how much I enjoy this discussion! 

> Option (3a) is: “This document describes the latest version of the spec. You MUST support that version as described here. You MAY support older versions if you want, it’s up to you. If you want to do that, please refer to the COMPATIBILITY WITH OLDER VERSIONS section below for details of what has changed in the last version and hints on how to deal with older versions.”

I am sold! Would you be interested to try and draft the basic guidelines to enacting your proposal? I think we should do a SSSOM 1.0 before the end of the year, which also explicitly introduces your (3a) suggestion. I do not think we should try (and I don't think you are suggesting this) to figure out the dozens of breaking changes that happened in the past. But it would be good if the guidelines are in place, and documented explicitly as part of the "spec", before we publish 1.0.

**Comment by gouttegd** [comment]
> what exactly would happen if an uncontrolled field (no fixed values) is changed from type string to type EntityRef?

Another option that I did not include yesterday, but that I think could the best option if the field is *really* important: phased deprecation.

That is, let’s say that in SSSOM version _X_ we have a _foo_ uncontrolled field that we wish to change to EntityRef with a controlled vocabulary; that field is too important to ignore but alas it is also impossible to define an automated way of mapping the thousands of free text values in use in the wild to the controlled vocabulary.

What we can do then:

* In SSSOM version _X+1_, we leave _foo_ unchanged and we introduce a new field called _foo_id_ that uses the controlled vocabulary. Both fields are present simultaneously in the schema, so all implementations MUST support both. If both fields are present in a mapping set, implementations MUST ignore _foo_ and only use _foo_id_. If only _foo_ is present, implementations MUST accept it as before but MUST print a warning that _foo_ is going to be deprecated and that SSSOM producers should use _foo_id_.
* In SSSOM version _X+2_, we remove _foo_ from the model. Implementations MUST support _foo_id_, but support for the uncontrolled _foo_ is now optional. If an implementation does decide to support _foo_, it should treat it as in version _X+1_: accept it but print a (stronger) warning that “_foo_ is now obsolete, you should really update your mapping sets to use _foo_id_, you know“.

We could even do a 3-phases deprecation, depending on how important the field is: in _X+2_, support for _foo_ is optional but recommended (implementations SHOULD support _foo_), in _X+3_ and ulterior support for _foo_ is fully optional (implementations MAY support _foo_).

Of course it’s complicated, cumbersome, but it may be the only way to ensure a smooth transition if/when we need an important breaking change post-1.0. (All the more reason to avoid such breaking changes whenever possible!)

**Comment by gouttegd** [comment]
> Would you be interested to try and draft the basic guidelines to enacting your proposal?

I opened the ticket in the first place, so I better be interested! :D

> I do not think we should try (and I don't think you are suggesting this) to figure out the dozens of breaking changes that happened in the past.

Agreed. I have used the past breaking changes as examples above since they were there, but I am not suggesting that SSSOM 1.0 should recommend supporting features from pre-1.0 versions, even less _all_ features from pre-1.0 versions.

I do think that if we go any of the (3) options, SSSOM 1.0 should document how to support at least one or two old fields that don’t exist anymore (e.g. `match_type`), because it will give implementers a good idea of what they could expect for possible future breaking changes.

Such support would be entirely optional though (MAY). For example, `sssom-java` does still support `match_type`, but `sssom-py` does not (it flatly rejects a mapping set that uses `match_type` instead of `mapping_justification`), and I think it’s **perfectly fine**.

**Comment by matentzn** [comment]
Wow.. I can't remember a more vigilant defender of adopters than you. Ok, make a proposal as a PR on the spec whenever you find the energy and time, and we has out the details. I like the idea of phased deprecation, but it will necessitate divorcing the TSV implementation entirely from LinkML, which is in any case necessary as per our discussion on serialisation versioning.

**Comment by gouttegd** [comment]
I am not sure there is anything left to do here. We now know how to deal with maintaining backwards compatibility, and version 1.0 of the spec already contains some examples of that.

What remains is the possibility of adding a `sssom_version` slot in future version, but if we do that I think this could be a distinct, focused issue – this one has seen a long enough discussion already.

**Comment by matentzn** [comment]
Awesome progress, I think yes, scanning through this this can be closed and left to the archeologists!
